[CI] Tests that are running at midnight and assert on log contents can fail because the log rolls over

There is a failure in https://elasticsearch-ci.elastic.co/job/elastic+elasticsearch+7.0+packaging-tests/306/console that took me a while to understand.

The error is this:

```
1) test42BundledJdkRemoved(org.elasticsearch.packaging.test.OssDebBasicTests)
java.lang.AssertionError: 
Expected: a string containing "/usr/lib/jvm/java-8-openjdk-amd64/jre"
     but: was "[2019-04-23T00:00:02,240][INFO ][o.e.d.DiscoveryModule    ] [packer-virtualbox-iso-1528442324] using discovery type [zen] and seed hosts providers [settings]
[2019-04-23T00:00:02,548][INFO ][o.e.n.Node               ] [packer-virtualbox-iso-1528442324] initialized
[2019-04-23T00:00:02,548][INFO ][o.e.n.Node               ] [packer-virtualbox-iso-1528442324] starting ...
[2019-04-23T00:00:02,636][INFO ][o.e.t.TransportService   ] [packer-virtualbox-iso-1528442324] publish_address {127.0.0.1:9300}, bound_addresses {[::1]:9300}, {127.0.0.1:9300}
[2019-04-23T00:00:02,643][WARN ][o.e.b.BootstrapChecks    ] [packer-virtualbox-iso-1528442324] the default discovery settings are unsuitable for production use; at least one of [discovery.seed_hosts, discovery.seed_providers, cluster.initial_master_nodes] must be configured
[2019-04-23T00:00:02,653][INFO ][o.e.c.c.ClusterBootstrapService] [packer-virtualbox-iso-1528442324] no discovery configuration found, will perform best-effort cluster bootstrapping after [3s] unless existing master is discovered
[2019-04-23T00:00:02,795][INFO ][o.e.c.s.MasterService    ] [packer-virtualbox-iso-1528442324] elected-as-master ([1] nodes joined)[{packer-virtualbox-iso-1528442324}{sjbbwqAqSlqxMWo74zo2PQ}{2Qe-mtCQSTKyp9nbqj1i_g}{127.0.0.1}{127.0.0.1:9300} elect leader, _BECOME_MASTER_TASK_, _FINISH_ELECTION_], term: 2, version: 7, reason: master node changed {previous [], current [{packer-virtualbox-iso-1528442324}{sjbbwqAqSlqxMWo74zo2PQ}{2Qe-mtCQSTKyp9nbqj1i_g}{127.0.0.1}{127.0.0.1:9300}]}
[2019-04-23T00:00:02,809][INFO ][o.e.c.s.ClusterApplierService] [packer-virtualbox-iso-1528442324] master node changed {previous [], current [{packer-virtualbox-iso-1528442324}{sjbbwqAqSlqxMWo74zo2PQ}{2Qe-mtCQSTKyp9nbqj1i_g}{127.0.0.1}{127.0.0.1:9300}]}, term: 2, version: 7, reason: Publication{term=2, version=7}
[2019-04-23T00:00:02,835][INFO ][o.e.g.GatewayService     ] [packer-virtualbox-iso-1528442324] recovered [0] indices into cluster_state
[2019-04-23T00:00:02,839][INFO ][o.e.h.AbstractHttpServerTransport] [packer-virtualbox-iso-1528442324] publish_address {127.0.0.1:9200}, bound_addresses {[::1]:9200}, {127.0.0.1:9200}
[2019-04-23T00:00:02,839][INFO ][o.e.n.Node               ] [packer-virtualbox-iso-1528442324] started
[2019-04-23T00:00:02,977][INFO ][o.e.c.m.MetaDataCreateIndexService] [packer-virtualbox-iso-1528442324] [library] creating index, cause [auto(bulk api)], templates [], shards [1]/[1], mappings []
[2019-04-23T00:00:03,229][INFO ][o.e.c.m.MetaDataMappingService] [packer-virtualbox-iso-1528442324] [library/KW8GCkWKS22L_-uo0JoCfw] create_mapping [book]
[2019-04-23T00:00:03,370][INFO ][o.e.c.m.MetaDataDeleteIndexService] [packer-virtualbox-iso-1528442324] [library/KW8GCkWKS22L_-uo0JoCfw] deleting index
[2019-04-23T00:00:04,186][INFO ][o.e.c.m.MetaDataCreateIndexService] [packer-virtualbox-iso-1528442324] [library] creating index, cause [auto(bulk api)], templates [], shards [1]/[1], mappings []
[2019-04-23T00:00:04,234][INFO ][o.e.c.m.MetaDataMappingService] [packer-virtualbox-iso-1528442324] [library/iQ_gggZtTm-K1SDFnelhVg] create_mapping [book]
[2019-04-23T00:00:04,278][INFO ][o.e.c.m.MetaDataDeleteIndexService] [packer-virtualbox-iso-1528442324] [library/iQ_gggZtTm-K1SDFnelhVg] deleting index
[2019-04-23T00:00:04,296][INFO ][o.e.n.Node               ] [packer-virtualbox-iso-1528442324] stopping ...
[2019-04-23T00:00:04,313][INFO ][o.e.n.Node               ] [packer-virtualbox-iso-1528442324] stopped
[2019-04-23T00:00:04,313][INFO ][o.e.n.Node               ] [packer-virtualbox-iso-1528442324] closing ...
[2019-04-23T00:00:04,321][INFO ][o.e.n.Node               ] [packer-virtualbox-iso-1528442324] closed
"
```

I assume it's trying to find `/usr/lib/jvm/java-8-openjdk-amd64/jre` in the log line that looks similar to this:

```
[2019-04-23T13:51:22,273][INFO ][o.e.n.Node               ] [node-0] JVM home [/usr/lib/jvm/java-8-openjdk-amd64/jre]
```

That log line isn't present at all in the failing log file, presumably because it was logged before midnight and hence when the test checks the contents of the latest log file shortly after midnight it's not in there but in the previous day's log file.

To be robust against log file rolling, tests that check for particular log messages could also check the previous day's logs.

--- a/qa/vagrant/src/main/java/org/elasticsearch/packaging/test/ArchiveTestCase.java
+++ b/qa/vagrant/src/main/java/org/elasticsearch/packaging/test/ArchiveTestCase.java
@@ -31,7 +31,6 @@ import org.elasticsearch.packaging.util.ServerUtils;
 import org.elasticsearch.packaging.util.Shell;
 import org.elasticsearch.packaging.util.Shell.Result;
 
-import java.nio.charset.StandardCharsets;
 import java.nio.file.Files;
 import java.nio.file.Path;
 import java.nio.file.Paths;
@@ -169,8 +168,8 @@ public abstract class ArchiveTestCase extends PackagingTestCase {
         Archives.stopElasticsearch(installation);
 
         String systemJavaHome = sh.getEnv().get("JAVA_HOME");
-        Path log = installation.logs.resolve("elasticsearch.log");
-        assertThat(new String(Files.readAllBytes(log), StandardCharsets.UTF_8), containsString(systemJavaHome));
+        assertThat(FileUtils.slurpAllLogs(installation.logs, "elasticsearch.log", "*.log.gz"),
+            containsString(systemJavaHome));
     }
 
     public void test51JavaHomeOverride() throws Exception {

--- a/qa/vagrant/src/main/java/org/elasticsearch/packaging/test/PackageTestCase.java
+++ b/qa/vagrant/src/main/java/org/elasticsearch/packaging/test/PackageTestCase.java
@@ -35,7 +35,6 @@ import java.nio.file.Paths;
 import java.nio.file.StandardOpenOption;
 import java.util.regex.Matcher;
 import java.util.regex.Pattern;
-import java.util.StringJoiner;
 
 import static com.carrotsearch.randomizedtesting.RandomizedTest.getRandom;
 import static org.elasticsearch.packaging.util.FileUtils.append;
@@ -125,15 +124,8 @@ public abstract class PackageTestCase extends PackagingTestCase {
             Files.write(installation.envFile, originalEnvFile);
         }
 
-        Path log = installation.logs.resolve("elasticsearch.log");
-        StringJoiner logFileJoiner = new StringJoiner("\n");
-        logFileJoiner.add(new String(Files.readAllBytes(log), StandardCharsets.UTF_8));
-
-        for (Path rotatedLogFile : FileUtils.lsGlob(installation.logs, "*.log.gz")) {
-            logFileJoiner.add(FileUtils.slurpGZ(rotatedLogFile));
-        }
-
-        assertThat(logFileJoiner.toString(), containsString(systemJavaHome));
+        assertThat(FileUtils.slurpAllLogs(installation.logs, "elasticsearch.log", "*.log.gz"),
+            containsString(systemJavaHome));
     }
 
     public void test32JavaHomeOverride() throws Exception {

--- a/qa/vagrant/src/main/java/org/elasticsearch/packaging/test/PackageTestCase.java
+++ b/qa/vagrant/src/main/java/org/elasticsearch/packaging/test/PackageTestCase.java
@@ -35,6 +35,7 @@ import java.nio.file.Paths;
 import java.nio.file.StandardOpenOption;
 import java.util.regex.Matcher;
 import java.util.regex.Pattern;
+import java.util.StringJoiner;
 
 import static com.carrotsearch.randomizedtesting.RandomizedTest.getRandom;
 import static org.elasticsearch.packaging.util.FileUtils.append;
@@ -125,7 +126,14 @@ public abstract class PackageTestCase extends PackagingTestCase {
         }
 
         Path log = installation.logs.resolve("elasticsearch.log");
-        assertThat(new String(Files.readAllBytes(log), StandardCharsets.UTF_8), containsString(systemJavaHome));
+        StringJoiner logFileJoiner = new StringJoiner("\n");
+        logFileJoiner.add(new String(Files.readAllBytes(log), StandardCharsets.UTF_8));
+
+        for (Path rotatedLogFile : FileUtils.lsGlob(installation.logs, "*.log.gz")) {
+            logFileJoiner.add(FileUtils.slurpGZ(rotatedLogFile));
+        }
+
+        assertThat(logFileJoiner.toString(), containsString(systemJavaHome));
     }
 
     public void test32JavaHomeOverride() throws Exception {

--- a/qa/vagrant/src/main/java/org/elasticsearch/packaging/util/FileUtils.java
+++ b/qa/vagrant/src/main/java/org/elasticsearch/packaging/util/FileUtils.java
@@ -25,8 +25,9 @@ import org.hamcrest.Matcher;
 
 import java.io.BufferedWriter;
 import java.io.ByteArrayOutputStream;
-import java.io.FileInputStream;
 import java.io.IOException;
+import java.nio.channels.Channels;
+import java.nio.channels.FileChannel;
 import java.nio.charset.StandardCharsets;
 import java.nio.file.DirectoryStream;
 import java.nio.file.Files;
@@ -129,7 +130,8 @@ public class FileUtils {
 
     public static String slurpGZ(Path file) {
         ByteArrayOutputStream fileBuffer = new ByteArrayOutputStream();
-        try (GZIPInputStream in = new GZIPInputStream(new FileInputStream(file.toString()))){
+        try (GZIPInputStream in = new GZIPInputStream(Channels.newInputStream(FileChannel.open(file))))
+        {
             byte[] buffer = new byte[1024];
             int len;
 

--- a/qa/vagrant/src/main/java/org/elasticsearch/packaging/util/FileUtils.java
+++ b/qa/vagrant/src/main/java/org/elasticsearch/packaging/util/FileUtils.java
@@ -130,6 +130,11 @@ public class FileUtils {
         }
     }
 
+    /**
+     * Returns the content a {@link java.nio.file.Path} file. The file can be in plain text or GZIP format.
+     * @param file The {@link java.nio.file.Path} to the file.
+     * @return The content of {@code file}.
+     */
     public static String slurpTxtorGz(Path file) {
         ByteArrayOutputStream fileBuffer = new ByteArrayOutputStream();
         try (GZIPInputStream in = new GZIPInputStream(Channels.newInputStream(FileChannel.open(file)))) {
@@ -151,6 +156,16 @@ public class FileUtils {
         }
     }
 
+    /**
+     * Returns combined content of a text log file and rotated log files matching a pattern. Order of rotated log files is
+     * not guaranteed.
+     * @param logPath Base directory where log files reside.
+     * @param activeLogFile The currently active log file. This file needs to be plain text under {@code logPath}.
+     * @param rotatedLogFilesGlob A glob pattern to match rotated log files under {@code logPath}.
+     *                            See {@link java.nio.file.FileSystem#getPathMatcher(String)} for glob examples.
+     * @return Merges contents of {@code activeLogFile} and contents of filenames matching {@code rotatedLogFilesGlob}.
+     * File contents are separated by a newline. The order of rotated log files matched by {@code rotatedLogFilesGlob} is not guaranteed.
+     */
     public static String slurpAllLogs(Path logPath, String activeLogFile, String rotatedLogFilesGlob) {
         StringJoiner logFileJoiner = new StringJoiner("\n");
         try {

--- a/qa/vagrant/src/main/java/org/elasticsearch/packaging/util/FileUtils.java
+++ b/qa/vagrant/src/main/java/org/elasticsearch/packaging/util/FileUtils.java
@@ -41,7 +41,9 @@ import java.util.ArrayList;
 import java.util.Arrays;
 import java.util.Collections;
 import java.util.List;
+import java.util.StringJoiner;
 import java.util.zip.GZIPInputStream;
+import java.util.zip.ZipException;
 
 import static org.hamcrest.MatcherAssert.assertThat;
 import static org.hamcrest.Matchers.emptyIterable;
@@ -128,18 +130,36 @@ public class FileUtils {
         }
     }
 
-    public static String slurpGZ(Path file) {
+    public static String slurpTxtorGz(Path file) {
         ByteArrayOutputStream fileBuffer = new ByteArrayOutputStream();
-        try (GZIPInputStream in = new GZIPInputStream(Channels.newInputStream(FileChannel.open(file))))
-        {
+        try (GZIPInputStream in = new GZIPInputStream(Channels.newInputStream(FileChannel.open(file)))) {
             byte[] buffer = new byte[1024];
             int len;
 
-            while((len = in.read(buffer)) != -1){
+            while ((len = in.read(buffer)) != -1) {
                 fileBuffer.write(buffer, 0, len);
             }
 
-            return(new String(fileBuffer.toByteArray(), StandardCharsets.UTF_8));
+            return (new String(fileBuffer.toByteArray(), StandardCharsets.UTF_8));
+        } catch (ZipException e) {
+            if (e.toString().contains("Not in GZIP format")) {
+                return slurp(file);
+            }
+            throw new RuntimeException(e);
+        } catch (IOException e) {
+            throw new RuntimeException(e);
+        }
+    }
+
+    public static String slurpAllLogs(Path logPath, String activeLogFile, String rotatedLogFilesGlob) {
+        StringJoiner logFileJoiner = new StringJoiner("\n");
+        try {
+            logFileJoiner.add(new String(Files.readAllBytes(logPath.resolve(activeLogFile)), StandardCharsets.UTF_8));
+
+            for (Path rotatedLogFile : FileUtils.lsGlob(logPath, rotatedLogFilesGlob)) {
+                logFileJoiner.add(FileUtils.slurpTxtorGz(rotatedLogFile));
+            }
+            return(logFileJoiner.toString());
         } catch (IOException e) {
             throw new RuntimeException(e);
         }

--- a/qa/vagrant/src/main/java/org/elasticsearch/packaging/util/FileUtils.java
+++ b/qa/vagrant/src/main/java/org/elasticsearch/packaging/util/FileUtils.java
@@ -24,6 +24,8 @@ import org.hamcrest.FeatureMatcher;
 import org.hamcrest.Matcher;
 
 import java.io.BufferedWriter;
+import java.io.ByteArrayOutputStream;
+import java.io.FileInputStream;
 import java.io.IOException;
 import java.nio.charset.StandardCharsets;
 import java.nio.file.DirectoryStream;
@@ -38,6 +40,7 @@ import java.util.ArrayList;
 import java.util.Arrays;
 import java.util.Collections;
 import java.util.List;
+import java.util.zip.GZIPInputStream;
 
 import static org.hamcrest.MatcherAssert.assertThat;
 import static org.hamcrest.Matchers.emptyIterable;
@@ -119,6 +122,22 @@ public class FileUtils {
     public static String slurp(Path file) {
         try {
             return String.join("\n", Files.readAllLines(file, StandardCharsets.UTF_8));
+        } catch (IOException e) {
+            throw new RuntimeException(e);
+        }
+    }
+
+    public static String slurpGZ(Path file) {
+        ByteArrayOutputStream fileBuffer = new ByteArrayOutputStream();
+        try (GZIPInputStream in = new GZIPInputStream(new FileInputStream(file.toString()))){
+            byte[] buffer = new byte[1024];
+            int len;
+
+            while((len = in.read(buffer)) != -1){
+                fileBuffer.write(buffer, 0, len);
+            }
+
+            return(new String(fileBuffer.toByteArray(), StandardCharsets.UTF_8));
         } catch (IOException e) {
             throw new RuntimeException(e);
         }

